src/tsolvers/egraph/Enode.h

```diff
@@ -59,154 +82,86 @@ struct ELRef {
-    ELRef getForbid     ()        const { assert(!isSymb()); if (isList()) return ELRef_Undef; else return ex->trm.forbid; }
-    ELRef& altForbid    ()              { assert(isTerm()); return ex->trm.forbid; }
-    void  setForbid     (ELRef r)       { assert(isTerm()); ex->trm.forbid = r; }
-    int   getDistIndex  ()        const { assert(!isSymb()); if (isList()) return 0; else return ex->trm.dist_index; }
-    void  setDistIndex  (int i)         { assert(isTerm()); ex->trm.dist_index = i; }
-    void  setDistClasses( const dist_t& d) { assert(!isSymb()); if (isList()) assert(d == 0); else ex->trm.dist_classes = d; }
-
-    inline dist_t getDistClasses() const { assert(!isSymb()); if (isTerm()) return ex->trm.dist_classes; else return 0; }
+
+    CgId  getCid() const { return cid; }
+    ERef getRoot() const { return root; }
+    void setRoot(ERef r) { root = r; }
+
+    ERef getEqNext () const {  return eq_next; }
+    void setEqNext (ERef n) { eq_next = n; }
+    int  getEqSize () const { return eq_size; }
+    void setEqSize (int i) { eq_size = i; }
+
+    PtAsgn getExpReason       () const { return exp_reason; }
+    ERef   getExpParent       () const { return exp_parent; }
+    ERef   getExpRoot         () const { return exp_root; }
+    int    getExpTimeStamp    () const { return exp_time_stamp; }
+
+    void setExpReason     (PtAsgn r)     { exp_reason = r; }
+    void setExpParent     (ERef r)       { exp_parent = r; }
+    void setExpRoot       (ERef r)       { exp_root   = r; }
+    void setExpTimeStamp  (const int t)  { exp_time_stamp = t; }
+
+    PTRef getTerm       ()        const { return pterm; }
+    ELRef getForbid     ()        const { return forbid; }
+    void  setForbid     (ELRef r)       { forbid = r; }
+    void  setDistClasses( const dist_t& d) { dist_classes = d; }
+    dist_t getDistClasses() const { return dist_classes; }
+
+    uint32_t getSize() const { return argSize; }
+    SymRef getSymbol() const { return symb; }
+    ERef operator[](std::size_t i) const { return *(args + i); }
+    ERef const * begin() const { return args; }
+    ERef const * end() const { return args + argSize; }
+    UseVectorIndex getIndex(uint32_t i) { return UseVectorIndex{static_cast<int32_t>((args + argSize + i)->x)}; }
+    void setIndex(uint32_t i, UseVectorIndex index) { (args + argSize + i)->x = static_cast<uint32_t>(index.x); }
 };
 
 struct ERefHash {
     uint32_t operator () (const ERef& s) const {
         return (uint32_t)s.x; }
 };
 
-struct ERef_vecHash {
-    uint32_t operator () (const vec<ERef>& s) const {
-        int m = 0; for (int i = 0; i < s.size(); i++) m += s[i].x;
-        return m; }
-};
-
-struct ERef_vecEq {
-    bool operator () (const vec<ERef>& s1, const vec<ERef>& s2) const {
-        if (s1.size() != s2.size()) return false;
-        for (int i = 0; i < s1.size(); i++)
-            if (s1[i] != s2[i]) return false;
-        return true; }
-};
-
 class EnodeAllocator : public RegionAllocator<uint32_t>
```

Comment:
There's so much code duplication with these allocators.  At some point somebody should try to write them with a template.